hack/local-cmo.sh

```diff
@@ -1,28 +1,103 @@
 #!/usr/bin/env bash
+set -eu -o pipefail
+
 # Script will:
 # 1. Patch CVO to exclude CMO from being managed
 # 2. Scale in-cluster CMO to 0
 # 3. Try to run CMO locally (CMO needs to be built before starting)
 
-set -eu
 
-# Go to the root of the repo
-cd "$(git rev-parse --show-cdup)"
+validate() {
+  local ret=0
+
+  make build || {
```

Comment:
Running `make build` will regenerate everything in `assets/`. This shouldn't be a part of this script.

/hold

Reply:
Please see: https://github.com/openshift/cluster-monitoring-operator/pull/1181#discussion_r639670188 

--- 
@paulfantom, IMHO, it makes sense to ensure that the operator runs fine and runs with the latest changes. Thus, having make build would reduce the edit, run, debug to just

edit the code
(re)run the script.
E.g.

fd .go$ | entr ./hack/local-cmo.sh
Power users are free to add/make changes to script. We could even add a NO_BUILD=true ./hack/local-cmo.sh

Reply:
@paulfantom could you please elaborate why assets shouldn't be regenerated?

Reply:
There are cases during development where we need to figure out how the stack will behave after applying some change to generated kubernetes manifests stored in `assets/`. It is much easier to change those yaml files manually, run the operator (with those changed files) and later tweak jsonnet, than to start figuring out which part of jsonnet code I need to change just to fix something. To put it in concrete example:
Let's say we got a bug about an incorrect label selector for a ServiceMonitor. It would be easy to:
1) start openshift cluster from cluster-bot
2) amend YAML file
3) build operator
4) run this script
5) amend the file again to add more things
6) DO NOT REBUILD OPERATOR as it is not needed because `assets/` are picked in runtime
7) run script again
8) repeat 5-7 until issue is fixed
9) find how to apply this in jsonnet (at this point cluster is no longer needed)
10) create PR or PRs when fix is also necessary in upstream projects

However if we include `make build` in this script, developer would need to find how to apply the fix in jsonnet first and regenerate `assets/` before every run of this script. Especially consider what happens when jsonnet code is vendored from external library like kube-prometheus.

Reply:
Thanks a lot for taking time to explain 🙇